src/main/scala/firrtl/Utils.scala

```diff
@@ -473,153 +383,139 @@ object Utils extends LazyLogging {
-        case e:Mux => MALE
-        case e:ValidIf => MALE
-        case e:WInvalid => MALE
-        case e => println(e); error("Shouldn't be here")
-    }}
-   def get_gender (s:Statement) : Gender =
-     s match {
-       case s:DefWire => BIGENDER
-       case s:DefRegister => BIGENDER
-       case s:WDefInstance => MALE
-       case s:DefNode => MALE
-       case s:DefInstance => MALE
-       case s:DefMemory => MALE
-       case s:Block => UNKNOWNGENDER
-       case s:Connect => UNKNOWNGENDER
-       case s:PartialConnect => UNKNOWNGENDER
-       case s:Stop => UNKNOWNGENDER
-       case s:Print => UNKNOWNGENDER
-       case EmptyStmt => UNKNOWNGENDER
-       case s:IsInvalid => UNKNOWNGENDER
+   def info (s:Statement) : Info = s match {
+      case s:DefWire => s.info
+      case s:DefRegister => s.info
+      case s:DefInstance => s.info
+      case s:WDefInstance => s.info
+      case s:DefMemory => s.info
+      case s:DefNode => s.info
+      case s:Conditionally => s.info
+      case s:PartialConnect => s.info
+      case s:Connect => s.info
+      case s:IsInvalid => s.info
+      case s:Stop => s.info
+      case s:Print => s.info
+      case s:Block => NoInfo
+      case EmptyStmt => NoInfo
+   }
+   def gender (e:Expression) : Gender = e match {
+     case e:WRef => e.gender
+     case e:WSubField => e.gender
+     case e:WSubIndex => e.gender
+     case e:WSubAccess => e.gender
+     case e:DoPrim => MALE
+     case e:UIntLiteral => MALE
+     case e:SIntLiteral => MALE
+     case e:Mux => MALE
+     case e:ValidIf => MALE
+     case e:WInvalid => MALE
+     case e => println(e); error("Shouldn't be here")
+   }
+   def get_gender (s:Statement) : Gender = s match {
+     case s:DefWire => BIGENDER
+     case s:DefRegister => BIGENDER
+     case s:WDefInstance => MALE
+     case s:DefNode => MALE
+     case s:DefInstance => MALE
+     case s:DefMemory => MALE
+     case s:Block => UNKNOWNGENDER
+     case s:Connect => UNKNOWNGENDER
+     case s:PartialConnect => UNKNOWNGENDER
+     case s:Stop => UNKNOWNGENDER
+     case s:Print => UNKNOWNGENDER
+     case EmptyStmt => UNKNOWNGENDER
+     case s:IsInvalid => UNKNOWNGENDER
+   }
+   def get_gender (p:Port) : Gender = if (p.direction == Input) MALE else FEMALE
+   def kind (e:Expression) : Kind = e match {
+     case e:WRef => e.kind
+     case e:WSubField => kind(e.exp)
+     case e:WSubIndex => kind(e.exp)
+     case e => ExpKind()
+   }
+   def tpe (e:Expression) : Type = e match {
+     case e:Reference => e.tpe
+     case e:SubField => e.tpe
+     case e:SubIndex => e.tpe
+     case e:SubAccess => e.tpe
+     case e:WRef => e.tpe
+     case e:WSubField => e.tpe
+     case e:WSubIndex => e.tpe
+     case e:WSubAccess => e.tpe
+     case e:DoPrim => e.tpe
+     case e:Mux => e.tpe
+     case e:ValidIf => e.tpe
+     case e:UIntLiteral => UIntType(e.width)
+     case e:SIntLiteral => SIntType(e.width)
+     case e:WVoid => UnknownType
+     case e:WInvalid => UnknownType
+   }
```

Comment:
While there is no hard and fast rule here, I think for some of these utility functions like info for Statements and tpe for both Statements and Expressions, we should delete the function and use the class methods. Same with get_type below.

For others like gender that is not currently a class method and perhaps shouldn't be, I think this solution is fine.